lib/mongodb.js

```diff
@@ -1996,23 +1996,97 @@ function optimizedFindOrCreate(model, filter, data, options, callback) {
 }
 
 /**
- * Convert the decimal properties from string to decimal.
+ * Convert decimal properties from string to decimal.
  * Only supported after mongodb 3.4
  *
  * @param {Object} data The data that might contain a decimal property
  * @param {Object} props The model property definitions
  */
 function convertDecimalProps(data, props) {
-  if (debug.enabled) debug('convertDecimalProps props: ', util.inspect(props));
-  for (const p in props) {
-    const prop = props[p];
-    const isDecimal = data[p] && prop && prop.mongodb &&
-      prop.mongodb.dataType &&
-      prop.mongodb.dataType.toLowerCase() === 'decimal128';
-    if (isDecimal) {
-      data[p] = Decimal128.fromString(data[p]);
-      debug('convertDecimalProps decimal value: ', data[p]);
+  visitAllProperties(data, props, coerceDecimalProperties);
+
+  /**
+ *
+ * @param {*} data Plain Data Object for the matching property definition(s)
+ * @param {*} propDefMap Property definition(s) which include information about property type
+ * @param {*} visitor A callback function which takes a property value and
+ * definition to apply custom property coercion
+ */
+  function visitAllProperties(data, propDefMap, visitor) {
```

Comment:
IMPORTANT❗️ 

IIUC, these new functions are defined inside `convertDecimalProps`. Since they are not (or at least should not) require anything from `convertDecimalProps` closure, they should be defined at the top-level, so that there is only one ("singleton") instance used by all invocations of `convertDecimalProps`. Such solution allows V8 to optimize performance.

Personally, I'd prefer to move this particular helper to `loopback-connector`, so that connectors for other databases can use it too. To land this pull request sooner, I think it will be best to make that change (move) after this patch is landed.

Reply:
>IIUC, these new functions are defined inside convertDecimalProps. Since they are not (or at least should not) require anything from convertDecimalProps closure, they should be defined at the top-level, so that there is only one ("singleton") instance used by all invocations of convertDecimalProps. Such solution allows V8 to optimize performance.

Good to know, thanks! That makes a lot of sense. Yea we can think of moving this function out to `loopback-connector` as discussed after this patch is landed.